File: services/api/memory.py
# Client intelligence layer: persistent profile store with provenance.
# Seed file is immutable; runtime file accumulates live facts. Reset = delete runtime.
import json
import random
import re
import string
import time

from data import API_DIR, iso_now

BASE36 = string.digits + string.ascii_lowercase


def _seed_path(client_id):
    return API_DIR / "memory" / f"{client_id}.seed.json"


def _runtime_path(client_id):
    return API_DIR / "memory" / f"{client_id}.runtime.json"


def _new_id(prefix):
    return f"{prefix}_{int(time.time() * 1000)}_{''.join(random.choices(BASE36, k=4))}"


def load_profile(client_id):
    runtime, seed_file = _runtime_path(client_id), _seed_path(client_id)
    if runtime.exists():
        profile = json.loads(runtime.read_text())
    elif seed_file.exists():
        profile = json.loads(seed_file.read_text())
    else:
        profile = {"clientId": client_id, "facts": [], "episodes": []}
    # Seed facts predate fact ids; assign stable ids on first load.
    assigned = False
    for i, f in enumerate(profile["facts"]):
        if not f.get("id"):
            f["id"] = f"fact_seed_{i}"
            assigned = True
    if assigned or not runtime.exists():
        _save(client_id, profile)
    return profile


def _save(client_id, profile):
    _runtime_path(client_id).write_text(json.dumps(profile, indent=2))


def _tokens(s):
    cleaned = re.sub(r"[^a-z0-9$ ]", "", s.lower())
    return {w for w in cleaned.split() if len(w) > 2}


def _similar(a, b):
    ta, tb = _tokens(a), _tokens(b)
    denom = min(len(ta), len(tb) or 1)
    if denom == 0:
        return False
    return len(ta & tb) / denom > 0.6


def add_facts(client_id, facts, episode_id=None):
    profile = load_profile(client_id)
    added = []
    for f in facts:
        if not f.get("fact") or not f.get("category"):
            continue
        existing = next(
            (x for x in profile["facts"] if x["status"] == "active" and _similar(x["fact"], f["fact"])),
            None,
        )
        # Identical restatement: keep the original (provenance points at first mention).
        if existing and existing["fact"] == f["fact"]:
            continue
        # Similar but different wording in the same category: newer fact supersedes.
        if existing and existing["category"] == f["category"]:
            existing["status"] = "superseded"
            existing["superseded_at"] = iso_now()
        elif existing:
            continue
        entry = {
            "id": _new_id("fact"),
            "fact": f["fact"],
            "category": f["category"],
            "source_episode_id": episode_id,
            "source_quote": f.get("source_quote", ""),
            "learned_at": iso_now(),
            "confidence": f.get("confidence", 0.8),
            "status": "active",
        }
        profile["facts"].append(entry)
        added.append(entry)
    if added:
        _save(client_id, profile)
    return added


# Governed deletion: facts are never removed, only marked deleted (append-only audit).
def forget_fact(client_id, fact_id):
    profile = load_profile(client_id)
    fact = next((f for f in profile["facts"] if f["id"] == fact_id and f["status"] == "active"), None)
    if not fact:
        return None
    fact["status"] = "deleted"
    fact["deleted_at"] = iso_now()
    _save(client_id, profile)
    return fact


def append_episode(client_id, session, role, content):
    profile = load_profile(client_id)
    ep = {"id": _new_id("ep"), "session": session, "role": role, "content": content, "timestamp": iso_now()}
    profile["episodes"].append(ep)
    _save(client_id, profile)
    return ep


def episodes_for(client_id, session):
    return [e for e in load_profile(client_id)["episodes"] if e["session"] == session]


def active_facts(client_id):
    return [f for f in load_profile(client_id)["facts"] if f["status"] == "active"]


def reset_profile(client_id):
    runtime = _runtime_path(client_id)
    if runtime.exists():
        runtime.unlink()
    return load_profile(client_id)  # re-seeds from .seed.json


def profile_as_context(client_id):
    facts = active_facts(client_id)
    if not facts:
        return "No profile facts learned yet."
    by_cat: dict[str, list] = {}
    for f in facts:
        by_cat.setdefault(f["category"], []).append(f)
    return "\n".join(
        f"{cat}:\n" + "\n".join(f"  - {f['fact']} (learned {f['learned_at'][:10]})" for f in fs)
        for cat, fs in by_cat.items()
    )

File: app/backend/lib/memory.js
// Client intelligence layer: persistent profile store with provenance.
// Seed file is immutable; runtime file accumulates live facts. Reset = delete runtime.
import { readFileSync, writeFileSync, existsSync, unlinkSync } from "node:fs";
import { join } from "node:path";
import { BACKEND_DIR } from "./data.js";

const seedPath = (id) => join(BACKEND_DIR, "memory", `${id}.seed.json`);
const runtimePath = (id) => join(BACKEND_DIR, "memory", `${id}.runtime.json`);

export function loadProfile(clientId) {
  let profile;
  if (existsSync(runtimePath(clientId))) {
    profile = JSON.parse(readFileSync(runtimePath(clientId), "utf8"));
  } else if (existsSync(seedPath(clientId))) {
    profile = JSON.parse(readFileSync(seedPath(clientId), "utf8"));
  } else {
    profile = { clientId, facts: [], episodes: [] };
  }
  // Seed facts predate fact ids; assign stable ids on first load.
  let assigned = false;
  profile.facts.forEach((f, i) => {
    if (!f.id) {
      f.id = `fact_seed_${i}`;
      assigned = true;
    }
  });
  if (assigned || !existsSync(runtimePath(clientId))) save(clientId, profile);
  return profile;
}

function save(clientId, profile) {
  writeFileSync(runtimePath(clientId), JSON.stringify(profile, null, 2));
}

const tokens = (s) => new Set(s.toLowerCase().replace(/[^a-z0-9$ ]/g, "").split(/\s+/).filter((w) => w.length > 2));
function similar(a, b) {
  const ta = tokens(a), tb = tokens(b);
  let inter = 0;
  for (const t of ta) if (tb.has(t)) inter++;
  return inter / Math.min(ta.size, tb.size || 1) > 0.6;
}

export function addFacts(clientId, facts, episodeId) {
  const profile = loadProfile(clientId);
  const added = [];
  for (const f of facts) {
    if (!f.fact || !f.category) continue;
    const existing = profile.facts.find((x) => x.status === "active" && similar(x.fact, f.fact));
    // Identical restatement: keep the original (provenance points at first mention).
    if (existing && existing.fact === f.fact) continue;
    // Similar but different wording in the same category: newer fact supersedes.
    if (existing && existing.category === f.category) {
      existing.status = "superseded";
      existing.superseded_at = new Date().toISOString();
    } else if (existing) {
      continue;
    }
    const entry = {
      id: `fact_${Date.now()}_${Math.random().toString(36).slice(2, 6)}`,
      fact: f.fact,
      category: f.category,
      source_episode_id: episodeId ?? null,
      source_quote: f.source_quote ?? "",
      learned_at: new Date().toISOString(),
      confidence: f.confidence ?? 0.8,
      status: "active",
    };
    profile.facts.push(entry);
    added.push(entry);
  }
  if (added.length) save(clientId, profile);
  return added;
}

// Governed deletion: facts are never removed, only marked deleted (append-only audit).
export function forgetFact(clientId, factId) {
  const profile = loadProfile(clientId);
  const fact = profile.facts.find((f) => f.id === factId && f.status === "active");
  if (!fact) return null;
  fact.status = "deleted";
  fact.deleted_at = new Date().toISOString();
  save(clientId, profile);
  return fact;
}

export function appendEpisode(clientId, { session, role, content }) {
  const profile = loadProfile(clientId);
  const ep = {
    id: `ep_${Date.now()}_${Math.random().toString(36).slice(2, 6)}`,
    session, role, content,
    timestamp: new Date().toISOString(),
  };
  profile.episodes.push(ep);
  save(clientId, profile);
  return ep;
}

export function episodesFor(clientId, session) {
  return loadProfile(clientId).episodes.filter((e) => e.session === session);
}

export function activeFacts(clientId) {
  return loadProfile(clientId).facts.filter((f) => f.status === "active");
}

export function resetProfile(clientId) {
  if (existsSync(runtimePath(clientId))) unlinkSync(runtimePath(clientId));
  return loadProfile(clientId); // re-seeds from .seed.json
}

export function profileAsContext(clientId) {
  const facts = activeFacts(clientId);
  if (!facts.length) return "No profile facts learned yet.";
  const byCat = {};
  for (const f of facts) (byCat[f.category] ??= []).push(f);
  return Object.entries(byCat)
    .map(([cat, fs]) => `${cat}:\n` + fs.map((f) => `  - ${f.fact} (learned ${f.learned_at.slice(0, 10)})`).join("\n"))
    .join("\n");
}
